src/Orleans/Utils/Pool/ConcurrentObjectPool.cs

```diff
@@ -0,0 +1,37 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Threading;
+
+namespace Orleans.Runtime
+{
+    public class ConcurrentObjectPool<T> : IObjectPool<T>
+        where T : PooledResource<T>, IDisposable
+    {
+        private readonly ConcurrentBag<T> _objects;
+        private readonly Func<T> _objectGenerator;
+
+        public ConcurrentObjectPool(Func<T> objectGenerator)
```

Comment:
I'm planning to adjust the pool implementation after initial perf test results. As for pre-allocating - client side pool uses way less messages than silo, so most of preallocated messages will remain unused. And in trade off between client memory consumption + startup times vs more aligned gen2 layout, i prefer the former. But if you certain that it's worth it - I'll make this change.